feat: introduce download proxy

XREvo opened the pull request:
This PR relates to Issue https://github.com/boring-registry/boring-registry/issues/180

This is still a work in progress (even it's fully functionnal).

Merging shoudl only be considered when an agreement will be acquired on the Issue https://github.com/boring-registry/boring-registry/issues/180

oliviermichaelis reviewed (commented):
Thanks for your work @XREvo!

I reviewed the code and left some final comments :)

I haven't tested the changes properly yet. Once I find the time, I'll make sure to test it with the remaining storage providers

XREvo commented on cmd/root.go:
```diff
@@ -70,6 +73,7 @@ func Execute() {
 func init() {
 	rootCmd.PersistentFlags().BoolVar(&flagJSON, "json", false, "Enable json logging")
 	rootCmd.PersistentFlags().BoolVar(&flagDebug, "debug", false, "Enable debug logging")
+	rootCmd.PersistentFlags().BoolVar(&flagProxy, "proxy", false, "Enable proxying download request to remote storage")
```
You are right!

I didn't get it in mind when adding it here. I'll move it to `serverCmd` and rename it as suggested

XREvo commented on pkg/core/proxifier.go:
```diff
@@ -0,0 +1,84 @@
+
+// Get the root part of the URL of the request
+func getRootURLFromRequest(r *http.Request) string {
+	// Find the protocol (http ou https)
```
French reflex...sorry

XREvo commented on pkg/module/transport.go:
```diff
@@ -37,6 +37,7 @@ func MakeHandler(svc Service, auth endpoint.Middleware, metrics *o11y.ModuleMetr
 				append(
 					options,
 					httptransport.ServerBefore(extractMuxVars(varNamespace, varName, varProvider)),
+					httptransport.ServerBefore(core.ExtractRootUrl()),
```
Since listing do not return URL, it is not needed (it's a mistake of mine)

XREvo commented on pkg/proxy/endpoint.go:
```diff
@@ -0,0 +1,90 @@
+	return func(ctx context.Context, request interface{}) (interface{}, error) {
+		input := request.(proxyRequest)
+
+		metrics.Download.With(prometheus.Labels{}).Inc()
```
After testing it, I see that I can't code it by simply removing the `With(prometheus.Labels{})` part :
```bash
❯ go build
# github.com/boring-registry/boring-registry/pkg/proxy
pkg/proxy/endpoint.go:31:20: metrics.Download.Inc undefined (type *prometheus.CounterVec has no field or method Inc)
```
I must at least specify it with `nil` value for labels

oliviermichaelis replied:
You're right, let's leave it like it is. I think I recently spent too much time with other languages where the Prometheus library has a different API :)

oliviermichaelis reviewed (changes requested):
Just one small typo left

XREvo commented:
Everything has been fixed

oliviermichaelis commented:
How did you get it to work on your machine?
I tried it with S3 and `--download-proxy`, only to get 404s:
```shell
terraform init

Initializing the backend...
Initializing modules...
Downloading <redacted>:5601/hashicorp/tls-private-key/aws 0.1.1 for test...
╷
│ Error: Failed to download module
│
│   on main.tf line 1:
│    1: module "test" {
│
│ Could not download module "test" (main.tf:1) source code from
│ "https://<redacted>:5601/v1/proxy/modules%2Fhashicorp%2Ftls-private-key%2Faws%2Fhashicorp-tls-private-key-aws-0.1.1.tar.gz%3FX-Amz-Algorithm%3DAWS4-HMAC-SHA256%26X-Amz-Credential%<redacted>%252Feu-central-1%252Fs3%252Faws4_request%26X-Amz-Date%3D20240701T223153Z%26X-Amz-Expires%3D300%26X-Amz-SignedHeaders%<redacted>":
│ error downloading
│ 'https://<redacted>:5601/v1/proxy/modules%2Fhashicorp%2Ftls-private-key%2Faws%2Fhashicorp-tls-private-key-aws-0.1.1.tar.gz%3FX-Amz-Algorithm%3DAWS4-HMAC-SHA256%26X-Amz-Credential%<redacted>%252Feu-central-1%252Fs3%252Faws4_request%26X-Amz-Date%3D20240701T223153Z%26X-Amz-Expires%3D300%26X-Amz-SignedHeaders%<redacted>':
│ bad response code: 404.
╵
```

It must have something to do with the route handler not matching the `/v1/proxy/{url}` :thinking:

XREvo commented:
I've just re-tested everything (an e2e test with our current internal modules) and found bugs.

i've pushed the fixes:
- i've deleted the URL encoding, which was causing issues
- i've rewrote the handler's pattern and the url extractor for the proxy

Right now it works on my machine and on our test server

oliviermichaelis commented:
Some good news: I tested it for S3, MinIO, and GCS and it works perfectly! However, it doesn't work for Azure unfortunately.

The blob storage URLs are of the following format:
```
https://<storageAccount>.blob.core.windows.net/<storageContainer>/providers%2Fhashicorp%2Frandom%2Fterraform-provider-random_2.0.0_linux_amd64.zip?se=2024-07-11T00%3A32%3A30Z&sig=<redacted>&ske=2024-07-11T02%3A27%3A30Z&skoid=<redacted>&sks=b&skt=2024-07-10T22%3A27%3A30Z&sktid=<redacted>&skv=2023-11-03&sp=r&spr=https&sr=b&sv=2023-11-03
```

I think the `/proxy` endpoint unescapes the URL which results in the following `url` muxVar:
```
boring-registry/providers/hashicorp/random/terraform-provider-random_2.0.0_linux_amd64.zip\
```

The resulting download URL looks like this:
```
https://<storageAccount>.blob.core.windows.net//<storageContainer>/providers/hashicorp/random/terraform-provider-random_2.0.0_linux_amd64.zip\?se\=2024-07-11T00%3A32%3A30Z\&sig\=<redacted>\&ske\=2024-07-11T02%3A27%3A30Z\&skoid\=<redacted>\&sks\=b\&skt\=2024-07-10T22%3A27%3A30Z\&sktid\=<redacted>\&skv\=2023-11-03\&sp\=r\&spr\=https\&sr\=b\&sv\=2023-11-03
```

XREvo commented:
One of my colleagues gave me an access to a Storage Account on his tenant, in order to debug the situation.

I created a directory structure to host a Terraform module:  
![image](https://github.com/user-attachments/assets/d16b419d-126a-4a74-a570-13203514a8f1)

Which allowed me to download the module via the proxy URL:  
![image](https://github.com/user-attachments/assets/e964a016-6cf4-4510-a996-0de2c9e2ae83)

The bug was in the method used to rebuild the original storage URL. Since I didn't known how client URL would be formed, I implemented an string concatenation that was invalid

It works locally for me, can you test it again?

oliviermichaelis reviewed (approved):
I just tested it, everything works now :tada: 

Thank you for your contribution, I really appreciate it!

XREvo commented:
Thank you @oliviermichaelis  
It's a pleasure to work with you

And if you need help to implement anything else, just ping me
